From c9f57f7d1c34d664d382a80cbe8ed1ab8b904316 Mon Sep 17 00:00:00 2001
From: Xiao-Long Chen <chenxiaolong@cxl.epac.to>
Date: Sun, 25 Aug 2013 22:28:41 -0400
Subject: [PATCH] vold: Allow ASEC containers on external SD when internal is
 emulated

* This modifies Volume::mountVol() and Volume::unmountVol to call
  createBindMounts() to set up .android_secure/ regardless if the
  volume is primary or not. On devices where emulated internal
  storage is used, the emulated storage is primary. Without this
  patch, /data/secure/asec/ and the tmpfs over .android_secure/
  won't be mounted.

Thanks to Pierre-Hugues Husson for the idea in this patch! The initial
implementation called createBindMounts() on all mount points regardless
of whether or not it was the SD card.

Change-Id: Ic2c45d275179099065a10e1e664a08efea66acf1
---
 Volume.cpp        | 14 ++++++++++----
 Volume.h          |  1 +
 VolumeManager.cpp |  4 ++--
 3 files changed, 13 insertions(+), 6 deletions(-)

diff --git a/Volume.cpp b/Volume.cpp
index a7ac7f0..d3e627c 100644
--- a/Volume.cpp
+++ b/Volume.cpp
@@ -239,6 +239,11 @@ bool Volume::isPrimaryStorage() {
     return !strcmp(getMountpoint(), externalStorage);
 }
 
+bool Volume::isExternalSd() {
+    const char* externalSd = getenv("EXTERNAL_SD") ? : "/storage/sdcard1";
+    return !strcmp(getMountpoint(), externalSd);
+}
+
 void Volume::setLunNumber(int lunNumber) {
     mLunNumber = lunNumber;
 }
@@ -396,6 +401,7 @@ int Volume::mountVol() {
     int n, i, rc = 0;
     char errmsg[255];
     bool primaryStorage = isPrimaryStorage();
+    bool externalSd = isExternalSd();
     char decrypt_state[PROPERTY_VALUE_MAX];
     char crypto_state[PROPERTY_VALUE_MAX];
     char encrypt_progress[PROPERTY_VALUE_MAX];
@@ -597,8 +603,8 @@ int Volume::mountVol() {
 
         protectFromAutorunStupidity();
 
-        // only create android_secure on primary storage
-        if (primaryStorage && createBindMounts()) {
+        // only create android_secure on external SD
+        if (externalSd && createBindMounts()) {
             SLOGE("Failed to create bindmounts (%s)", strerror(errno));
             umount("/mnt/secure/staging");
             setState(Volume::State_Idle);
@@ -849,8 +855,8 @@ int Volume::unmountVol(bool force, bool revert) {
     setState(Volume::State_Unmounting);
     usleep(1000 * 1000); // Give the framework some time to react
 
-    /* Undo createBindMounts(), which is only called for primary storage */
-    if (isPrimaryStorage()) {
+    /* Undo createBindMounts() */
+    if (isExternalSd()) {
         /*
          * Remove the bindmount we were using to keep a reference to
          * the previously obscured directory.
diff --git a/Volume.h b/Volume.h
index d1f053f..311fae2 100644
--- a/Volume.h
+++ b/Volume.h
@@ -78,6 +78,7 @@ public:
     const char *getMountpoint() { return mMountpoint; }
     int getState() { return mState; }
     bool isPrimaryStorage();
+    bool isExternalSd();
 
     int getLunNumber() { return mLunNumber; }
     void setLunNumber(int lunNumber);
diff --git a/VolumeManager.cpp b/VolumeManager.cpp
index 8a74872..5bd2cb5 100644
--- a/VolumeManager.cpp
+++ b/VolumeManager.cpp
@@ -1600,8 +1600,8 @@ bool VolumeManager::isMountpointMounted(const char *mp)
 }
 
 int VolumeManager::cleanupAsec(Volume *v, bool force) {
-    /* Only EXTERNAL_STORAGE needs ASEC cleanup. */
-    if (!v->isPrimaryStorage())
+    /* Only EXTERNAL_SD needs ASEC cleanup. */
+    if (!v->isExternalSd())
         return 0;
 
     int rc = unmountAllAsecsInDir(Volume::SEC_ASECDIR_EXT);
-- 
1.8.4

